server/src/main/java/org/elasticsearch/search/vectors/ExactKnnQueryBuilder.java

```diff
@@ -66,14 +84,24 @@ public String getWriteableName() {
 
     @Override
     protected void doWriteTo(StreamOutput out) throws IOException {
-        out.writeFloatArray(query);
+        if (out.getTransportVersion().onOrAfter(TransportVersions.KNN_EXPLICIT_BYTE_QUERY_VECTOR_PARSING)) {
+            boolean isFloat = query.isFloatVector();
+            out.writeBoolean(isFloat);
+            if (isFloat) {
+                out.writeFloatArray(query.asFloatVector());
+            } else {
+                out.writeByteArray(query.asByteVector());
+            }
+        } else {
+            out.writeFloatArray(query.asFloatVector());
+        }
         out.writeString(field);
     }
 
     @Override
     protected void doXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject(NAME);
-        builder.field("query", query);
+        builder.field("query", query.asFloatVector());
```

Comment:
Transforming `byte -> float` here isn't necessary. The `VectorData` should determine which "kind" it is and not bother transforming between them. Encapsulating the toXContent should fix this.